### PR TITLE
Add slide cropping

### PR DESCRIPTION
**Pull request by Philbrick**
- Add slide cropping to crop downsamples to create uniformity in pixel spacing across pyramid levels
- Adds end to end tests

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -140,7 +141,14 @@ int main(int argc, char *argv[]) {
         ("compressionQuality",
         programOptions::value<int>(
         &compressionQuality)->default_value(80),
-        "Compression quality range(0 - 100");
+        "Compression quality range(0 - 100")
+        ("uniformPixelSpacing",
+        programOptions::bool_switch(
+        &cropFrameToGenerateUniformPixelSpacing)->default_value(false),
+        "Crop imaging to generate downsampled mages with unifrom pixel "
```
Comment (danielbeaudreau) [inline src/main.cpp]: typo

File: src/wsiToDcm.cpp [hunk 2]
```diff
@@ -172,6 +220,7 @@ std::unique_ptr<SlideLevelDim> WsiToDcm::getSlideLevelDim(const int32_t level,
       downsampleOfLevel = static_cast<double>(downsample) / multiplicator;
       // check that downsampling is going from higher to lower magnification
       if (downsampleOfLevel >= 1.0) {
+        levelToGet = priorLevel->level;
```
Comment (danielbeaudreau) [inline src/wsiToDcm.cpp:248]: Why do we need to set levelToGet, levelWidth, or levelHeight here when they get set below? (line 233)

Just wondering if you've tested the scenario where we specify downsamples in the wrong order (decreasing).
Seems like we would just downsample based on the top level for each downsampled level.

File: src/wsiToDcm.cpp [hunk 3]
```diff
@@ -252,70 +312,102 @@ double  WsiToDcm::getDownsampledLevelDimensionMM(
...
+
+  BOOST_LOG_TRIVIAL(debug) << "Cropping source image: " <<
+                              firstSlideWidthCrop <<
+                              ", " <<
+                              firstSlideHeightCrop;
+  if (firstSlideWidthCrop >= firstLevelWidth_ ||
+      firstSlideHeightCrop >= firstLevelHeight_) {
+      BOOST_LOG_TRIVIAL(info) << "Error Pixel spacing normalization " <<
+                         "cropped entire input image. Cannot generate DICOM.";
+      return 0;
```
Comment (danielbeaudreau) [inline src/wsiToDcm.cpp]: should it be return 1 for an error

File: src/wsiToDcm.cpp [hunk 4]
```diff
@@ -252,70 +312,102 @@ double  WsiToDcm::getDownsampledLevelDimensionMM(
...
+    smallestSlideDim = NULL;
+  }
+
+  BOOST_LOG_TRIVIAL(debug) << "Cropping source image: " <<
+                              firstSlideWidthCrop <<
+                              ", " <<
+                              firstSlideHeightCrop;
+  if (firstSlideWidthCrop >= firstLevelWidth_ ||
+      firstSlideHeightCrop >= firstLevelHeight_) {
+      BOOST_LOG_TRIVIAL(info) << "Error Pixel spacing normalization " <<
```
Comment (danielbeaudreau) [inline src/wsiToDcm.cpp]: Is this condition possible?

File: src/wsiToDcm.cpp [hunk 5]
```diff
@@ -252,70 +312,102 @@ double  WsiToDcm::getDownsampledLevelDimensionMM(
...
+  // slide pixel spacing normalization croping to ensure pixel
+  // spacing across all downsample images is uniform.
+  int64_t firstSlideWidthCrop;
+  int64_t firstSlideHeightCrop;
+  std::vector<int32_t> slideLevels;
+  std::unique_ptr<SlideLevelDim> smallestSlideDim = std::move(
+                                          getSmallestSlideDim(&slideLevels));
+  if (wsiRequest_->cropFrameToGenerateUniformPixelSpacing) {
+    std::unique_ptr<SlideLevelDim> firstSlideLevel =
+                  std::move(getSlideLevelDim(0, NULL, smallestSlideDim.get()));
```
Comment (danielbeaudreau) [inline src/wsiToDcm.cpp]: Is the slide at index 0 always guaranteed to be the largest?

File: src/wsiToDcm.cpp [hunk 6]
```diff
@@ -472,12 +568,14 @@ int WsiToDcm::dicomizeTiff() {
     }
     higherMagnifcationDicomFiles.set_dicom_files(
                                            std::move(generatedDicomFiles));
+    // The combination of cropFrameToGenerateUniformPixelSpacing &
+    // can result in cropping images which span multiple frames/times
```
Comment (danielbeaudreau) [inline src/wsiToDcm.cpp]: nit: comment seems incomplete, combination of cropping and _?

File: tests/wsiToDcm.cpp [hunk 7]
```diff
@@ -136,3 +136,252 @@ TEST(compressionString, unknown) {
...
+  ASSERT_EQ(slide_dim->multiplicator, 1);
+  ASSERT_EQ(slide_dim->downsampleOfLevel, 2);
+  ASSERT_EQ(slide_dim->levelWidth, 2220);
+  ASSERT_EQ(slide_dim->levelHeight, 2967);
+  ASSERT_EQ(slide_dim->levelFrameWidth, 100);
+  ASSERT_EQ(slide_dim->levelFrameHeight, 100);
+  ASSERT_EQ(slide_dim->frameWidthDownsampled, 200);
+  ASSERT_EQ(slide_dim->frameHeightDownsampled, 200);
+  ASSERT_EQ(slide_dim->frameWidthDownsampled, 200);
+  ASSERT_EQ(slide_dim->frameHeightDownsampled, 200);
```
Comment (danielbeaudreau) [inline tests/wsiToDcm.cpp]: nit: assertion is done twice

File: tests/wsiToDcm.cpp [hunk 8]
```diff
@@ -136,3 +136,252 @@ TEST(compressionString, unknown) {
...
+  ASSERT_EQ(slide_dim->cropSourceLevelWidth, 2220 - 2208);
+  ASSERT_EQ(slide_dim->cropSourceLevelHeight, 2967 - 2944);
+  ASSERT_TRUE(slide_dim->readOpenslide);
```
Comment (danielbeaudreau) [inline tests/wsiToDcm.cpp]: nit: could you add a comment explaining where 2208 and 2944 come from? e.g. 2208 = lowest dim (69) * largest downsample (32)

File: src/main.cpp (hunk 1 above)
Comment (Philbrick) [reply src/main.cpp]: fixed

File: src/wsiToDcm.cpp (hunk 3 above)
Comment (Philbrick) [reply src/wsiToDcm.cpp]: fixed

File: tests/wsiToDcm.cpp (hunk 7 above)
Comment (Philbrick) [reply tests/wsiToDcm.cpp]: removed duplicates

File: tests/wsiToDcm.cpp (hunk 8 above)
Comment (Philbrick) [reply tests/wsiToDcm.cpp]: Added.

File: src/wsiToDcm.cpp (hunk 6 above)
Comment (Philbrick) [reply src/wsiToDcm.cpp]: fixed.

File: src/wsiToDcm.cpp (hunk 5 above)
Comment (Philbrick) [reply src/wsiToDcm.cpp]: No, great catch.

File: src/wsiToDcm.cpp (hunk 4 above)
Comment (Philbrick) [reply src/wsiToDcm.cpp]: It shouldn't be.  removing.

File: src/wsiToDcm.cpp (hunk 5 above)
Comment (Philbrick) [reply src/wsiToDcm.cpp]: Fixed.

File: src/wsiToDcm.cpp (hunk 2 above)
Comment (Philbrick) [reply src/wsiToDcm.cpp:248]: levelToGet and the other parameters are set here when the slide is being progressive downsampled.  generateUsingOpenSlide = false so the next if loop won't be entered.  Some of these are not actually used if progressive downsampling is being used. I set them here largely for completeness to a avoid undefined weirdness.

Yes, your comment, about level zero made me realize that the downsampling list enabled downsamples to be ordered in any order.  This would have broken the code.  The most recent CL should address this edge case and add unit tests to test it. Its not something that my usage of the code needs. But, should make it complete for others.

File: src/wsiToDcm.cpp [hunk 9]
```diff
@@ -329,33 +371,95 @@ std::unique_ptr<SlideLevelDim>  WsiToDcm::getSmallestSlideDim(
...
-                                   smallestSlideDim->levelHeightDownsampled;
-      slideLevels->push_back(level);
+                                   tempSlideLevelDim->levelHeightDownsampled;
+      if ((smallestSlideDim == NULL) ||
+          ((!wsiRequest_->stopDownsamplingAtSingleFrame ||
+            !smallest_level_is_single_frame) &&
+           (temp_downsample > smallsest_downsample)) ||
+          (smallest_level_is_single_frame &&
+           wsiRequest_->stopDownsamplingAtSingleFrame &&
+           frameCount == 1 && temp_downsample < smallsest_downsample)) {
```
Comment (danielbeaudreau) [inline src/wsiToDcm.cpp]: nit: this condition is hard to read. could you pull out the compound conditions as bool values defined above and then use those in the if statement argument

File: src/wsiToDcm.cpp [hunk 10]
```diff
@@ -321,6 +359,10 @@ std::unique_ptr<SlideLevelDim>  WsiToDcm::getSmallestSlideDim(
   } else {
     levels = svsLevelCount_;
   }
+  bool smallest_level_is_single_frame = false;
+  std::vector<std::tuple<int32_t, int64_t>> levelProcessOrder;
+  int64_t smallsest_downsample;
```
Comment (danielbeaudreau) [inline src/wsiToDcm.cpp]: typo

File: src/wsiToDcm.cpp [hunk 11]
```diff
@@ -312,6 +336,20 @@ double  WsiToDcm::getDownsampledLevelDimensionMM(
   return levelDim * firstLevelMpp / 1000;
 }
 
+bool downsample_order(const std::tuple<int32_t, int64_t> &i,
+                        const std::tuple<int32_t, int64_t> &j) {
+  const int64_t iDownsample = std::get<1>(i);
+  const int64_t jDownsample = std::get<1>(j);
+  if (iDownsample != jDownsample) {
+    // Sorta Downsampel : e/g/. Area 1, 2, 4, 8. 16
```
Comment (danielbeaudreau) [inline src/wsiToDcm.cpp]: typo

File: src/wsiToDcm.cpp [hunk 12]
```diff
@@ -329,33 +371,95 @@ std::unique_ptr<SlideLevelDim>  WsiToDcm::getSmallestSlideDim(
...
+           (temp_downsample > smallsest_downsample)) ||
+          (smallest_level_is_single_frame &&
+           wsiRequest_->stopDownsamplingAtSingleFrame &&
+           frameCount == 1 && temp_downsample < smallsest_downsample)) {
+        // if (smallest slice level not initalized) or
+        //
+        // if ((not stopDownsamplingAtSingleFrame or haven't seen single
+        // frame yet) and dimensions are smaller than previous frame) or
+        //
+        // if smallest area is single frame a stop downsampling at single
```
Comment (danielbeaudreau) [inline src/wsiToDcm.cpp]: typo: "a" should be "and"?

File: src/wsiToDcm.cpp [hunk 13]
```diff
@@ -329,33 +371,95 @@ std::unique_ptr<SlideLevelDim>  WsiToDcm::getSmallestSlideDim(
...
-            static_cast<double>(smallestSlideDim->levelFrameWidth));
-        const int64_t frameY = std::ceil(
-            static_cast<double>(smallestSlideDim->levelHeightDownsampled) /
-            static_cast<double>(smallestSlideDim->levelFrameHeight));
         BOOST_LOG_TRIVIAL(debug) << "Level[" <<level << "] frames:" <<
                                     frameX << ", " << frameY;
-        if (frameX <= 1 && frameY <= 1) {
+        if (frameCount <= 1) {
+          smallest_level_is_single_frame = true;
+        }
```
Comment (danielbeaudreau) [inline src/wsiToDcm.cpp]: nit: not sure if the encasing if statement is necessary here

File: src/wsiToDcm.cpp [hunk 14]
```diff
@@ -98,14 +97,34 @@ WsiToDcm::WsiToDcm(WsiRequest *wsiRequest) : wsiRequest_(wsiRequest) {
   }
   svsLevelCount_ = openslide_get_level_count(osr_);
   openslide_get_level_dimensions(osr_, 0,
```
Comment (danielbeaudreau) [inline src/wsiToDcm.cpp:100]: should this be -1 too?

File: tests/wsiToDcm.cpp [hunk 15]
```diff
@@ -385,3 +450,67 @@ TEST(getSlideLevelDim, cropping_progressive) {
...
+  request.downsamples = std::move(downsamples);
+  request.debug = true;
+  wsiToDicomConverter::WsiToDcm converter(&request);
+  std::vector<int32_t> levels;
+  std::unique_ptr<wsiToDicomConverter::SlideLevelDim> slide_dim;
+  std::unique_ptr<wsiToDicomConverter::SlideLevelDim> smallestSlideDim;
+  smallestSlideDim = std::move(converter.getSmallestSlideDim(&levels));
+  ASSERT_EQ(levels.size(), 6);
+  // -1 = get largest slice.
+  ASSERT_EQ(smallestSlideDim->levelWidthDownsampled, 69);
```
Comment (danielbeaudreau) [inline tests/wsiToDcm.cpp]: nit: should we add an assertion for level values similar to the other test case?

**Review by danielbeaudreau — CHANGES_REQUESTED** [review]

File: src/wsiToDcm.cpp (hunk 9 above)
Comment (Philbrick) [reply src/wsiToDcm.cpp]: Done.

File: src/wsiToDcm.cpp (hunk 10 above)
Comment (Philbrick) [reply src/wsiToDcm.cpp]: Done.

File: src/wsiToDcm.cpp (hunk 12 above)
Comment (Philbrick) [reply src/wsiToDcm.cpp]: Done.

File: src/wsiToDcm.cpp (hunk 13 above)
Comment (Philbrick) [reply src/wsiToDcm.cpp]: Moved up a level.  Done.

File: tests/wsiToDcm.cpp (hunk 15 above)
Comment (Philbrick) [reply tests/wsiToDcm.cpp]: Yes.  Done.

File: src/wsiToDcm.cpp (hunk 14 above)
Comment (Philbrick) [reply src/wsiToDcm.cpp:100]: No.  Its a Openslide API call.  Added comment.

File: src/wsiToDcm.cpp [hunk 16]
```diff
@@ -252,70 +337,183 @@ double  WsiToDcm::getDownsampledLevelDimensionMM(
...
+      if (setSmallestSlice) {
+        smallestSlideDim = std::move(tempSlideLevelDim);
+        smallestDownsample = tempDownsample;
+        BOOST_LOG_TRIVIAL(debug) << "Set Smallest";
+      }
+      std::tuple<int32_t, int64_t> tpl(level, tempDownsample);
+      levelProcessOrder.push_back(tpl);
+      if (wsiRequest_->stopDownsamplingAtSingleFrame && frameCount <= 1) {
+        BOOST_LOG_TRIVIAL(debug) << "Level[" <<level << "] frames:" <<
+                                    frameX << ", " << frameY;
```
Comment (danielbeaudreau) [inline src/wsiToDcm.cpp]: nit: this debug log message isn't very useful as it will just log for the one frame with 1 x 1 dimension. Should we move it out of the if statement?

**Review by danielbeaudreau — CHANGES_REQUESTED** [review]

File: src/wsiToDcm.cpp (hunk 16 above)
Comment (Philbrick) [reply src/wsiToDcm.cpp]: done

**Review by danielbeaudreau — APPROVED** [review]
